models/user.rb

```diff
@@ -232,14 +233,32 @@ def replace_username_in_all_content(new_username)
     end
   end
 
-
   def mark_as_read(thread)
     reconnect_mongo_primary
     read_state = read_states.find_or_create_by(course_id: thread.course_id)
     read_state.last_read_times[thread.id.to_s] = Time.now.utc
     read_state.save
   end
 
+  ##
+  # Update stats for a the user for the specified course by the specified amounts.
+  # The amounts here are increments/decrements, and not absolute
+  def update_stats_for_course(
+    course_id, threads: nil, responses: nil, replies: nil, active_flags: nil, inactive_flags: nil
+  )
+    begin
+      stats = course_stats.find(course_id: course_id)
+      stats.inc(threads: threads) if threads
+      stats.inc(responses: responses) if responses
+      stats.inc(replies: replies) if replies
+      stats.inc(active_flags: active_flags) if active_flags
+      stats.inc(inactive_flags: inactive_flags) if inactive_flags
+    rescue Mongoid::Errors::DocumentNotFound
+      # If the stats don't already exist, rebuild them from scratch, in which case they will have the correct counts.
+      build_course_stats_for_user(self, course_id)
```

Comment:
Since this rebuild happens as a trigger from posting/deleting comments/threads, can't this give unexpected slow responses, or even timeouts, the first time the stats are generated in big courses?
Maybe it'd be better to pre-generate stats as a migration step, instead of sneaking it into the `after_*` hooks.

Reply:
Oh, I just saw the comment about this being scoped to the user, so this should be fine. Nevermind this.

Reply:
I did consider making this a background task, however, for a single user and course, it should be a pretty small query, even for a very active user.